### PR TITLE
ListView shouldn't be scrollable when there isn't any scroll extent

### LINKED ISSUE #8271: Expansion panel is (unexpectedly) scrollable and clips
There are a few other demos which seem to have this (regression?) behavior.

expansion mov

### LINKED ISSUE #8276: bottom sheet in scrolling list demo difficult to dismiss
It should be possible to pull down from the top of the persistent bottom sheet to make it disappear, but it seems to have a very small hitbox to make that happen?

hard_dismiss mov

### LINKED ISSUE #8278: progress indicators demo is scrollable. :/
This seems like a regression from the scrolling refactor.  Possibly the same as https://github.com/flutter/flutter/issues/8271.

progress_indicator mov

### PR DESCRIPTION
**Pull request by abarth**
Previously, a ListView would always accept user input, even if it wasn't
actually scrollable. Now, by default, we don't accept user input if there's no
scroll range. You can override this behavior using the ScrollPhysics.

Fixes #8276
Fixes #8278
Fixes #8271

### CODE REVIEW

**Comment by abarth** [comment]
@HansMuller @Hixie

File: packages/flutter/lib/src/widgets/scroll_physics.dart [hunk 1]
```diff
@@ -145,6 +145,28 @@ class ClampingScrollPhysics extends ScrollPhysics {
...
+///
+/// On Android, overscrolls will be clamped by default and result in an
+/// overscroll glow. On iOS, overscrolls will load a spring that will return
+/// the scroll view to its normal range when released.
+///
+/// See also:
+///
+/// * [BouncingScrollPhysics], which provides the bouncing overscroll behavior
+///   found on iOS.
+/// * [ClampingScrollPhysics], which provides the claming overscroll behavior
```
Comment (HansMuller) [inline packages/flutter/lib/src/widgets/scroll_physics.dart]: clam => clamp

File: packages/flutter/lib/src/widgets/scroll_physics.dart [hunk 2]
```diff
@@ -145,6 +145,28 @@ class ClampingScrollPhysics extends ScrollPhysics {
...
+///   foudn on Android.
```
Comment (HansMuller) [inline packages/flutter/lib/src/widgets/scroll_physics.dart]: found

File: packages/flutter/lib/src/widgets/scroll_position.dart [hunk 3]
```diff
@@ -51,6 +51,17 @@ abstract class ScrollPhysics {
     return parent.applyPhysicsToUserOffset(position, offset);
   }
 
+  /// Whether the scrollable should let the user adjust the scroll offset, for
+  /// example by dragging.
+  ///
+  /// By default, the user can manipulate the scroll offset if there is some
+  /// difference between the min and max scroll extent.
```
Comment (HansMuller) [inline packages/flutter/lib/src/widgets/scroll_position.dart]: This is certainly correct however it might be worth reiterating the point in layman's terms. Something like: [in other words], if it's not possible for the user to expose more of the scrollable by tapping or dragging, then return false.

**Comment by HansMuller** [comment]
LGTM

**Comment by Hixie** [comment]
I think this was a bit over-aggressive. Lists should be able to overscroll even if they have no content, it's only things like dialog box bodies that only overscroll in "emergencies" that should turn off scrolling entirely like this.

**Comment by abarth** [comment]
This fixed a lot of interaction bugs in apps.  If we loosen up here, we probably need to implement scroll chaining to prevent "dead" areas of non-scrolling.

**Comment by realcarlos** [comment]
This will make problem when set physics: const BouncingScrollPhysics() in the ListView ,the pull refresh will not work if the ListView has no content ,because it is unscrollable.

**Comment by realcarlos** [comment]
We at least should keep the bounce switch

**Comment by Hixie** [comment]
You can override the physics on your list view to bounce even when it's empty. Just provide an AlwaysScrollableScrollPhysics instance as your physics object.
